src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs

```diff
@@ -3,40 +3,245 @@
+                    |                        Error Code (32)                        |
+                    +---------------------------------------------------------------+
+                */
+                case Http2FrameType.RST_STREAM:
+                    frame.RstStreamErrorCode = (Http2ErrorCode)BinaryPrimitives.ReadUInt32BigEndian(buffer);
+                    break;
+
+                /* https://tools.ietf.org/html/rfc7540#section-6.9
+                    +-+-------------------------------------------------------------+
+                    |R|              Window Size Increment (31)                     |
+                    +-+-------------------------------------------------------------+
+                */
+                case Http2FrameType.WINDOW_UPDATE:
+                    frame.WindowUpdateSizeIncrement = (int)Bitshifter.ReadUInt31BigEndian(buffer);
+                    break;
+
+                case Http2FrameType.PING: // Opaque payload 8 bytes long
+                case Http2FrameType.SETTINGS: // Settings are general payload
+                case Http2FrameType.CONTINUATION: // None
+                case Http2FrameType.PUSH_PROMISE: // Not implemented frames are ignored at this phase
+                default:
+                    return 0;
+            }
+
+            return extendedHeaderLength;
+        }
+
+        // The length in bytes of additional fields stored in the payload section.
+        // This may be variable based on flags, but should be no more than 8 bytes.
+        public int GetPayloadFieldsLength(Http2Frame frame)
+        {
+            switch (frame.Type)
+            {
+                // TODO: Extract constants
+                case Http2FrameType.DATA: // Variable 0 or 1
+                    return frame.DataHasPadding ? 1 : 0;
+                case Http2FrameType.HEADERS:
+                    return (frame.HeadersHasPadding ? 1 : 0) + (frame.HeadersHasPriority ? 5 : 0); // Variable 0 to 6
+                case Http2FrameType.GOAWAY:
+                    return 8; // Last stream id and error code.
+                case Http2FrameType.PRIORITY:
+                    return 5; // Stream dependency and weight
+                case Http2FrameType.RST_STREAM:
+                    return 4; // Error code
+                case Http2FrameType.WINDOW_UPDATE:
+                    return 4; // Update size
+                case Http2FrameType.PING: // 8 bytes of opaque data
+                case Http2FrameType.SETTINGS: // Settings are general payload
+                case Http2FrameType.CONTINUATION: // None
+                case Http2FrameType.PUSH_PROMISE: // Not implemented frames are ignored at this phase
+                default:
+                    return 0;
+            }
+        }
+
+        public IList<Http2PeerSetting> ReadSettings(ReadOnlySequence<byte> payload)
+        {
+            var data = payload.ToArray().AsSpan();
```

Comment:
This is super inefficient. There's a GetSpan helper in Kestrel that only allocates if it needs to. Use that.